### ISSUE TITLE
async generator literals using await are incorrectly resolved to Generator

### ISSUE THREAD

**Issue by FasterSpeeding** [microsoft/pyright#2322]
**Describe the bug**
Async generator literals made in the style of of `(await callback(value) for value in ...)` (as opposed to using async for) are incorrectly resolved to `Generator[ReturnT, None, None]` leading to false-positive errors from pyright when trying to treat the generator as an AsyncGenerator.

**To Reproduce**
Run pyright over the provided test code and watch it produce the following output

```
Searching for source files
Found 1 source file
C:\Users\Lucina\PycharmProjects\hikari-sake\test.py
  C:\Users\Lucina\PycharmProjects\hikari-sake\test.py:13:20 - error: "Generator[int, None, None]" is not iterable
    "__aiter__" method not defined (reportGeneralTypeIssues)
  C:\Users\Lucina\PycharmProjects\hikari-sake\test.py:14:15 - warning: Argument type is unknown
    Argument corresponds to parameter "values" in function "print" (reportUnknownArgumentType)
1 error, 1 warning, 0 infos
Completed in 0.58sec
```
Then run the following code and watch it pass without error and output `<class 'async_generator'>` for `print(type(bar))`.

**Expected behavior**

Pyright should resolve `bar = (await foo() for _ in range(0, 10))` to `AsyncGenerator[int, None]` and not error on the example code.

**Screenshots or Code**
```python
import asyncio
import random


async def foo() -> int:
    await asyncio.sleep(0.3)
    return random.randint(0, 100)


async def main() -> None:
    bar = (await foo() for _ in range(0, 10))
    print(type(bar))
    async for i in bar:
        print(i)


asyncio.run(main())
```

**VS Code extension or command-line**
pyright 1.1.170 (as a command line tool)
Targeting python 3.9

**Additional context**
This appears to have been reported back in https://github.com/microsoft/pylance-release/issues/1348 /https://github.com/microsoft/pyright/issues/1912 but I also tried v1.1.145 and still got this erroneous behaviour (although the other issues brought up there around building async generators in a sync function did seem to have been resolved).

**Comment by erictraut** [comment]
I think the current behavior is correct. It matches the behavior of mypy.

You can override the inferred type for the generator by explicitly annotating the type of `bar`. This eliminates the error.

```python
async def main() -> None:
    bar: AsyncGenerator[int, None] = (await foo() for _ in range(0, 10))
    print(type(bar))
    async for i in bar:
        print(i)
```

**Comment by FasterSpeeding** [comment]
> it matches the behaviour of mypy 

For what it's worth https://github.com/python/mypy/issues/10534 (which also brings up the resolution of generators which cLl await) was opened on MyPy around the same time as those issues I mentioned in the addition context section but it seems to have been mostly ignored or missed

**Comment by FasterSpeeding** [comment]
Also just a quick question but is there any case where using await like that in a generator literal would lead to Generator rather than AsyncGenerator at runtime that you know of? I can't really find any myself but I might just be missing something obvious here

**Comment by erictraut** [comment]
Ah, I didn't know about that mypy issue. Thanks for pointing that out.

Are you suggesting that any generator with an `await` keyword should always be treated as async and any generator without an `await` should always be sync? I have to admit that I don't know much about how async generator statements are used in practice.

**Comment by FasterSpeeding** [comment]
> Are you suggesting that any generator with an await keyword should always be treated as async and any generator without an await should always be sync? I have to admit that I don't know much about how async generator statements are used in practice.

Not quite any generator with an await keyword in it; while an async generator is the result of the case my previous example covered where the await statement is in the ELT part of the generator expression, the following example (while technically having await within the generator) resolves to a normal Generator (I had a better poke around this behaviour today).

```py
import asyncio
import random
import typing

_T = typing.TypeVar("_T")

async def process(value: _T) -> _T:
    return value


async def main():
    print(type(foo for foo in await test()))

asyncio.run(main())

```

It's also worth noting that if there's an await statement in the "if" part of a generator comprehension then it'll also resolve to an async generator under CPython's reference implementation (e.g. `(foo for foo in [1, 2, 3] if await check(foo))`).

But yeah I do personally think that that type checkers should be more aware of this behaviour

**Comment by erictraut** [comment]
OK, does this summary look correct to you?

1. If a generator has an `await` in the final `for` target, it is considered async.
2. If a generate has an `await` in any `if` clause, it is considered async.
3. In other cases (even if an `await` is used in the `in` clause), it is considered sync.

**Comment by FasterSpeeding** [comment]
> OK, does this summary look correct to you?
> 
> 1. If a generator has an `await` in the final `for` target, it is considered async.
> 2. If a generate has an `await` in any `if` clause, it is considered async.
> 3. In other cases (even if an `await` is used in the `in` clause), it is considered sync.

Yeah that sounds correct to me

**Comment by erictraut** [comment]
Thanks for your help in clarifying the correct behavior.

This change will appear in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.171 which was just published. It will also be included in the next release of pylance.
